### PR TITLE
db/etl: read a buffer with Next instead of Get(i)

### PR DESCRIPTION
**Pull request by AskAlexSharov**
context: reducing amount of expensive `re-allocs` inside `SD.mu` critical section (part of `release/3.6` `_newPayload` perf jumps problem). Independent of #23599 — this is the same clean-up applied to the buffer `main` has today.

- **`Buffer.Get(i)` becomes `Next()`.** The one production caller, `memoryDataProvider`, walked `i` upward anyway; the index and its `currentIndex` both go. `Sort` puts the cursor at the first entry, so Sorting again re-reads.
- **`entryLoc` loses `insertionOrder`** — offsets already rise with insertion order, so the sort's tie-break uses one. `Put` reserves a byte for an entry with no bytes, which is what makes that true for those too: four all-empty entries used to share offset 0 and could come back in any order. 12 bytes an entry rather than 20.
- **`Write` no longer allocates.** `w.Write` takes an `io.Writer`, so the local length array escaped and cost an allocation on every call. It lives on the buffer now — 1 → 0 allocs/op.
- **Spill files use a fixed length prefix** in front of each key and value instead of a varint: 2 bytes for a key (`nilKeyLen` reserved), 4 for a value. They never leave the machine that wrote them, so the length reads back in one load instead of a decode loop.
- **`Collect` rejects a key past `MaxKeyLen` (4096) or a value past `MaxValLen`.** A key spells its length in 2 bytes with `nilKeyLen` reserved, so the hard ceiling is 65534 and 4096 is policy under it; a longer value would write a length that reads back negative and desync the file. Both bounds are exported so callers can check against them.

`binary.NativeEndian` is now used everywhere in the package that encodes a length; there was no other site.

The format trade, alternating rounds on an M4:

| | varint | fixed |
|---|---|---|
| `SortableBufferLoadOnly/random_500k` | 18.8ms | **16.7ms** |
| spill bytes per entry (32B key, 128B value) | 163 | 166 (+1.8%) |

`MergeSortFiles/file_only_100k` is ~1.5% slower and inside its own noise — it spills 64 small files, where the extra 3 bytes an entry cost more than the decode saves.

Also ports the benchmarks and tests that apply from the chunked-buffer work: `Write` and `PutOnlyCold`, the all-empty-entries and cross-chunk stability tests that pin the offset tie-break, an allocation-free read test, and `InmemLoadOnly` renamed to `Read` — which now sorts once before the loop instead of inside the first timed iteration.

### CODE REVIEW

File: db/etl/buffers.go [hunk 1]
```diff
@@ -48,26 +48,45 @@ const (
 )
```
Comment (awskii) [inline db/etl/buffers.go:48]: `entryLocSize` on the line above is still 16, and its comment still enumerates `insertionOrder(4)`, which this change deletes. `entryLoc` is now three int32 fields = 12 bytes, and nothing pins the constant to the struct — there is no `unsafe.Sizeof` assertion in the package.

`Size()` is the only consumer and it feeds `CheckFlushSize()`, so every collector over-accounts 4 bytes per entry and spills early. With 32-byte keys and 64-byte values that is 112 accounted against 108 actual, crossing `optimalSize` ~3.7% early. The `indexKeys` inverted-index collector (8-byte key, ~52-byte value, `SmallSortableBuffers` at 32MB) reports 76 against an actual 72 and flushes at ~94.7% of target. More temp files per collation and more merge providers than configured — the opposite direction from this PR's own "16 bytes an entry rather than 20".

File: db/etl/buffers.go [hunk 2]
```diff
@@ -48,26 +48,45 @@ const (
...
 
 // writeSortedEntries writes buffer entries to w in varint-length-prefixed format.
```
Comment (awskii) [inline db/etl/buffers.go]: This sentence is now false and attached to the wrong declaration.

`writeSortedEntries` no longer writes varints — it calls `putKeyLen`/`putValLen`, which write fixed 2- and 4-byte `binary.NativeEndian` lengths. And the new const block was inserted between this line and the function, so line 50 is now the godoc for `keyLenSize`/`valLenSize`/`nilKeyLen` while `writeSortedEntries` sits undocumented at line 72.

`go doc etl` renders the const block introduced by a sentence naming an unrelated function and describing the encoding this change removed. The next two lines of the same comment already say the format is fixed width *rather than* a varint, so the block contradicts itself.

File: db/etl/buffers.go [hunk 3]
```diff
@@ -48,26 +48,45 @@ const (
...
-func writeSortedEntries(w io.Writer, entries []sortableBufferEntry) error {
-	var numBuf [binary.MaxVarintLen64]byte
+// A spill file prefixes each field with its length. Fixed width rather than a
+// varint: the file never leaves the machine that wrote it. A key is capped at
+// maxKeyLen so two bytes hold it; a value has no cap.
+const (
+	keyLenSize = 2
+	valLenSize = 4
+	nilKeyLen  = math.MaxUint16 // no key reaches this, so it can mean nil
```
Comment (awskii) [inline db/etl/buffers.go:56]: The `nilKeyLen` sentinel has no end-to-end coverage. `putKeyLen` encodes it and `readKeyLen` (dataprovider.go:161-163) decodes it, and nothing exercises the pair through a spill file.

The nil-key tests (`TestSortableBufferNilAndEmptyKeys`, `TestSortableBufferAllEmptyEntries`) drain in memory via `drainBuffer` and never call `Write`. The tests that do write and read back — `TestWriteAndReadBufferEntry`, `TestSortableBufferOversizedEntry` — use only non-nil keys. `TestEmptyKeyValue` covers the nil *value* sentinel through a real spill, but its key is `[]byte{}`, not nil.

What survives: delete the `if n == nilKeyLen { return -1, nil }` branch from `readKeyLen`, or change `nilKeyLen` on one side only, and the suite stays green. A nil key then reads back as a 65535-byte field, `readAt` returns `io.ErrUnexpectedEOF`, and the load fails with "error while reading next element from disk".

No in-tree caller passes a nil key today, but `Buffer.Put` is exported and `writeSortedEntries` goes out of its way to handle `entry.key == nil`.

File: db/etl/buffers.go [hunk 4]
```diff
@@ -106,6 +125,9 @@ const (
 	// entryLoc.offset can pack the chunk index with the offset inside the chunk
 	// and splitting the two is a shift and a mask. 1MB is also the least a
 	// collector can hold once it takes a chunk at all.
+	// A key must fit a chunk, since Sort slices it straight out of one.
```
Comment (awskii) [inline db/etl/buffers.go]: The stated reason does not hold, and the real constraint is elsewhere.

`Put` computes `n := max(len(k)+len(v), 1)` and calls `nextChunk(n)`, which allocates a private chunk of exactly `n` bytes when `n > dataChunkSize`. The entry lands at `off == 0` and both `Sort`'s comparator and `entryData` slice it correctly — `TestSortableBufferOversizedEntry` exercises that with a `dataChunkSize+7` entry. An entry fits its chunk by construction, and `dataChunkSize` is 1MB anyway, so 4096 is not derived from it.

What actually forces a cap is `keyLenSize = 2` with `nilKeyLen = math.MaxUint16` reserved: the ceiling is 65534. Nothing in code couples the two constants, so only this comment steers the decision.

The failure it sets up: someone needing longer keys reads "must fit a chunk", sees 1MB, and raises `maxKeyLen` to 65536. A 65535-byte key reaches `putKeyLen`, which writes `nilKeyLen`; `readKeyLen` decodes -1, `readKeyField` returns a nil key, the value length is parsed out of the key's payload, and the rest of the spill file desynchronises — `mergeSortFiles` loads garbage with no error.

```suggestion
	// keyLenSize is 2 and nilKeyLen reserves MaxUint16, so the ceiling is 65534.
```

Same rationale is repeated in the `TestCollectRejectsOversizedKey` docstring at etl_test.go:1802-1804.

**Review by awskii — COMMENTED** [review]
One point not anchored to a diff line.

`nextChunk`'s comment at buffers.go:231-232 still names `Get`: "An entry never straddles chunks, so Get can hand out direct references." `Buffer.Get(i)` is gone from every buffer type in this change. It is the only remaining `Get` reference in the package's non-test sources — `Allocator.Get` and `dataChunks.Get` are unrelated symbols — so the stale name reads as a live API rather than a dated one.

File: db/etl/buffers.go [hunk 5]
```diff
@@ -106,6 +125,9 @@ const (
...
+	maxKeyLen = 4096
```
Comment (awskii) [inline db/etl/buffers.go]: `maxKeyLen` was added as the first constant in the block, directly under the four-line comment explaining why the chunk size is a power of two and why 1MB. Go attaches that comment to `maxKeyLen`, leaving `dataChunkBits`/`dataChunkSize` — what it actually describes — bare below a blank line.

Hover or `go doc` on `maxKeyLen` returns the chunk-sizing rationale, which says nothing about the 4096 bound. Moving `maxKeyLen` below `dataChunkSize` with its own one-line comment fixes both this and the attachment of the existing block.

File: db/etl/buffers.go (hunk 1 above)
Comment (AskAlexSharov) [reply db/etl/buffers.go:48]: Fixed in 193ede9f03 — `entryLocSize` is 12, and the comment lists the three fields that are actually there.

Confirmed the accounting effect you describe: `Size()` is the only consumer and it feeds `CheckFlushSize()`, so every collector was charging 4 bytes an entry that no longer exist.

Added `TestEntryLocSize`, which asserts `unsafe.Sizeof(entryLoc{}) == entryLocSize`. Putting it back to 16 fails it — so the next field added or removed cannot drift the constant silently.

File: db/etl/buffers.go (hunk 2 above)
Comment (AskAlexSharov) [reply db/etl/buffers.go]: Fixed in 193ede9f03. The sentence naming `writeSortedEntries` moved back onto the function as `// writeSortedEntries writes the entries to w in the spill format above.`, and the const block keeps the paragraph that describes the format — which is what it documents.

Dropped "varint-length-prefixed" outright rather than correcting it, since the two lines below already say fixed width rather than a varint.

File: db/etl/buffers.go (hunk 3 above)
Comment (AskAlexSharov) [reply db/etl/buffers.go:56]: Fixed in 193ede9f03 — added `TestSpillNilKeyRoundTrip`. It Collects a nil key, calls `Flush` so `Load` reads a real file instead of RAM, and asserts the key comes back nil.

Ran your exact mutation to confirm it bites: deleting `if n == nilKeyLen { return -1, nil }` from `readKeyLen` fails the new test. Before it, the suite was green with that branch gone.

File: db/etl/buffers.go (hunk 4 above)
Comment (AskAlexSharov) [reply db/etl/buffers.go]: Fixed in 193ede9f03. You are right that the chunk is not what binds — an oversized entry gets a private chunk of exactly its size and lands at `off == 0`, so `Sort` slices it fine.

Took the substance of your suggestion rather than the line, because 4096 is not the ceiling either:

> A key spells its length in keyLenSize bytes with nilKeyLen reserved, so the ceiling is 65534. 4096 is policy under that ceiling - no caller comes near it, and Collect rejects the rest.

That way raising `maxKeyLen` reads as a policy change with a stated hard limit, instead of an invitation to go to 65536 and desynchronise the spill file the way you traced.

The same wrong rationale in `TestCollectRejectsOversizedKey` is updated too.

File: db/etl/buffers.go (hunk 5 above)
Comment (AskAlexSharov) [reply db/etl/buffers.go]: Fixed in 193ede9f03. `maxKeyLen` now sits below `dataChunkSize` with its own comment, so the power-of-two paragraph attaches to `dataChunkBits`/`dataChunkSize` — the declarations it describes.

**Review by awskii — APPROVED** [review]
Round 2 against 193ede9f. `entryLocSize` is 12 and pinned by a test, `maxKeyLen`'s rationale is corrected in the source, `writeSortedEntries` has its doc back.

One point has no changed line to anchor to. The PR body still describes 42c5c29, and this squash-merges, so that text becomes the permanent commit message for an on-disk format. Three of its claims no longer hold:

- "a fixed 4-byte length in front of each key and value" — the key prefix is 2 bytes (`keyLenSize = 2`, with `nilKeyLen` reserved). Only the value prefix is 4.
- "16 bytes an entry rather than 20" — it is 12 at this head (`entryLocSize = 12`, pinned by `TestEntryLocSize`).
- the trade-off row "168 (+3.1%)" is the 4+4 measurement. At 2+4 the overhead is 6 bytes: 166 (+1.8%).

It also still carries the "`Sort` slices a key straight out of a chunk, so it has to fit one" rationale that 193ede9f corrected at buffers.go:131-133.

Five minors follow as inline comments. None blocking.

File: db/etl/buffers.go [hunk 6]
```diff
@@ -44,30 +44,49 @@ const (
...
+
+func putValLen(dst []byte, valLen int32) {
+	binary.NativeEndian.PutUint32(dst, uint32(valLen)) //nolint:gosec
+}
+
+// writeSortedEntries writes the entries to w in the spill format above.
+func writeSortedEntries(w io.Writer, entries []sortableBufferEntry, numBuf []byte) error {
 	for _, entry := range entries {
-		lk := int64(len(entry.key))
+		keyLen, valLen := int32(len(entry.key)), int32(len(entry.value)) //nolint:gosec
```
Comment (awskii) [inline db/etl/buffers.go:91]: A value of 2 GiB or more decodes as nil here and desyncs the rest of the spill file.

`int32(len(entry.value))` goes negative past 2^31-1; `putValLen` writes it back as `uint32` (:67-68); `readValLen` decodes `int32(binary.NativeEndian.Uint32(...))` to a negative int (dataprovider.go:171); `readValField` returns nil on `n < 0` **without consuming the value bytes** (:197). Every read after that parses value payload as record headers.

The const comment at :51-52 says "a value has no cap", which is what makes this a defect rather than a documented limit. The old append/oldest spill path encoded `int64(len(value))`, so the ceiling is introduced here — only `sortableBuffer`'s own int32 limit was pre-existing.

Not reachable at any in-tree buffer threshold. Cheapest fix is the comment plus a guard in `Put`.

File: db/etl/collector.go [hunk 7]
```diff
@@ -93,7 +93,18 @@ func (c *Collector) SortAndFlushInBackground(v bool) *Collector {
 	return c
 }
 
+// errKeyTooLong is out of line so that building it does not cost the caller
```
Comment (awskii) [inline db/etl/collector.go]: The stated reason does not hold — `extractNextFunc` is not inlinable either way.

Go's budget is 80 cost units and a non-inlinable call costs 57. `extractNextFunc` makes at least six: `errKeyTooLong`, `c.allocator.Get()`, `getTypeByBuffer(c.buf)`, plus the interface dispatches `c.buf.Put`, `c.buf.CheckFlushSize()` and `c.flushBuffer(false)`. The two interface calls alone are 114, so it was over budget before this guard was added and there is no inlining budget for the hoist to protect.

No escape-analysis benefit either, which is the other reason this repo uses the pragma: the `[]any` `fmt.Errorf` builds escapes on that path only, and nothing in the frame is forced to the heap by it — `c.logPrefix` is an existing string field, `maxKeyLen` a constant.

Both other `//go:noinline` sites in non-test code carry rationales that do hold: db/seg/decompress.go:837 hoists a cold Huffman branch out of a genuinely small hot getter, execution/state/rw_v3.go:1515 keeps `&res` off the heap. This one matches the shape but not the substance.

`go build -gcflags=-m ./db/etl` settles it — I reasoned this from call count, not from a build.

**Review by awskii — COMMENTED** [review]
Two more with no changed line to anchor to.

`nextChunk`'s comment at buffers.go:234 still reads "so `Get` can hand out direct references". `Get` is gone from the interface and all three implementations, and every other `Get` in the package is unrelated — `Allocator.Get`, `dataChunks.Get`, `sync.Pool.Get` — so a reader chasing the name lands on a pool getter. The invariant it states is still real and still has consumers: `Next` (:311-318), `Write` (:384) and `Sort`'s comparator (:356-360) all slice straight out of one chunk via `entryData`. Same rename missed at etl_test.go:928, :938 and :1643 — the last on `TestSortableBufferOversizedEntry`, whose body this change rewrote to drop the `Get` calls it describes.

This is the one point from the last round with neither a fix nor a reply. It landed in a review body rather than a line comment, which is probably how it went missing.

Second: `Collect` gained a precondition its doc does not state. `extractNextFunc` now errors when `len(k) > maxKeyLen`, so the exported `Collect` has a new failure mode, its comment at collector.go:119 is unchanged, and `maxKeyLen` is unexported. Callers outside `db/etl` that build keys from variable-width data — `db/seg/compress.go:900`, `db/kv/prune/prune.go:110`, `db/state/inverted_index.go:982` — can neither read the bound from the signature nor reference the constant to check against it. Nothing in tree crosses it today (widest is `MaxPatternLen = 128`), so it is a contract gap rather than a live break.

File: db/etl/buffers.go (hunk 6 above)
Comment (AskAlexSharov) [reply db/etl/buffers.go:91]: Fixed in 66f4f3f0ac. Confirmed the mechanism you traced — the length reads back negative and `readValField` returns nil without consuming the bytes, so every later record is parsed out of value payload.

Guarded in two places rather than one, because a single `Put` is not the only way to get there: `Collect` rejects one oversized value, and `writeSortedEntries` rejects a value that `appendSortableBuffer.Put` grew past the bound across many Puts — which a `Put`-side check cannot see. Comment now reads "a key is capped at maxKeyLen so two bytes hold it, a value at maxValLen so four do".

`TestSpillValueLengthCeiling` pins the boundary without allocating 2 GiB: `maxValLen` survives the int32 round-trip and one more wraps negative.

File: db/etl/collector.go (hunk 7 above)
Comment (AskAlexSharov) [reply db/etl/collector.go]: You are right, and the build settles it your way. `go build -gcflags=-m ./db/etl` lists `(*Collector).Collect` as inlinable and does not list `extractNextFunc` — so the pragma was protecting a budget that was already gone.

Folded the helper back into `extractNextFunc` in 66f4f3f0ac rather than just correcting the comment: with no inlining or escape benefit there is nothing left for a separate function to do, and it is one less thing to explain.

**Review by awskii — APPROVED** [review]
